xformers/components/attention/csrc/cuda/attention.cu

```diff
@@ -1141,14 +1266,184 @@ std::tuple<at::Tensor, at::Tensor, at::Tensor> attention_backward(
+    int64_t end_iter = kBlockSizeK == 1 ? N : N - (N & (step - 1));
+    scalar_t s_delta[kBlockSizeQ][kBlockSizeK];
+    int64_t query_idx =
+        blockIdx.x * (blockDim.y * kBlockSizeQ) + threadIdx.y * kBlockSizeQ;
+    // if (l < end_iter) {
+    {
+      for (; l < end_iter; l += step) {
+        for (int jj = 0; jj < kBlockSizeQ; jj++) {
+          for (int kk = 0; kk < kBlockSizeK; kk++) {
+            s_delta[jj][kk] = 1;
+          }
+        }
+
+        apply_masking<scalar_t, vec_t, kBlockSizeK, kBlockSizeQ>(
+            s_delta, philox_args, global_offset, N, p, l);
+
+        for (int jj = 0; jj < kBlockSizeQ; jj++) {
+          for (int kk = 0; kk < kBlockSizeK; kk++) {
+            if (query_idx + jj < M)
+              output[jj][l + kk] = s_delta[jj][kk];
+          }
+        }
+      }
+    }
+    {
+      UnrollLoopForMask<
+          false,
+          scalar_t,
+          vec_t,
+          kBlockSizeK / 2,
+          kBlockSizeQ,
+          WARP_SIZE>::eval(output, N, M, philox_args, global_offset, p);
+    }
+  }
+};
+
+template <
+    bool first,
+    typename scalar_t,
+    typename vec_t,
+    int kBlockSizeQ,
+    int WARP_SIZE>
+struct UnrollLoopForMask<first, scalar_t, vec_t, 0, kBlockSizeQ, WARP_SIZE> {
+  static __device__ __forceinline__ void eval(
+      scalar_t* s_delta[kBlockSizeQ],
+      int64_t N,
+      int64_t M,
+      at::PhiloxCudaState philox_args,
+      int64_t global_offset,
+      scalar_t p) {}
+};
+
+template <
+    typename scalar_t,
+    typename vec_t,
+    int kBlockSizeK,
+    int kBlockSizeQ,
+    int WARP_SIZE>
+__global__ void dropout_kernel(
+    at::PackedTensorAccessor<scalar_t, 3> output,
+    scalar_t p,
+    at::PhiloxCudaState philox_args) {
+  static_assert(
+      integerIsPowerOf2(kBlockSizeK * WARP_SIZE),
+      "kBlockSizeK * WARP_SIZE should be a power of 2");
+  int64_t B = output.size(0);
+  int64_t M = output.size(1);
+  int64_t N = output.size(2);
+
+  int64_t batch_idx = blockIdx.y;
+  int64_t query_idx =
+      blockIdx.x * (blockDim.y * kBlockSizeQ) + threadIdx.y * kBlockSizeQ;
+
+  int64_t global_offset = batch_idx * M * N + query_idx * N;
+
+  if (query_idx >= M)
+    return;
+
+  scalar_t* output_block[kBlockSizeQ];
+  for (int64_t q_item_idx = 0; q_item_idx < kBlockSizeQ; q_item_idx++) {
+    int64_t index = query_idx + q_item_idx;
+    index = index >= M ? M - 1 : index;
+    output_block[q_item_idx] = output[batch_idx][index].data();
+  }
+
+  UnrollLoopForMask<
+      true,
+      scalar_t,
+      vec_t,
+      kBlockSizeK,
+      kBlockSizeQ,
+      WARP_SIZE>::eval(output_block, N, M, philox_args, global_offset, p);
+}
+
+at::Tensor _dropout_mask(at::Tensor output, double p) {
+  at::cuda::CUDAGuard device_guard(output.device());
+  int64_t B = output.size(0);
+  int64_t M = output.size(1);
+  int64_t N = output.size(2);
+
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+
+  constexpr int WARP_SIZE = 4;
```

Comment:
nit, but this looks like it's the number of warps, and warp_size is typically used (even in this PR, see next file) to describe the number of threads in a warp, right ? not super important but for consistency's sake

Reply:
Indeed. I'll clean this up in a follow-up PR, I have some changes layed out that I'll be pushing soon